tokio-postgres/tests/test/copy_both.rs

```diff
@@ -0,0 +1,91 @@
+use futures::{SinkExt, StreamExt, TryStreamExt};
+use std::future;
+use tokio_postgres::NoTls;
+use tokio_postgres::SimpleQueryMessage::Row;
+
+// https://stackoverflow.com/a/35907071
+fn contains(haystack: &[u8], needle: &[u8]) -> bool {
+    haystack
+        .windows(needle.len())
+        .any(|window| window == needle)
+}
+
+#[tokio::test]
+async fn copy_both() {
+    copy_both_impl(true).await;
+    copy_both_impl(false).await;
+}
+
+async fn copy_both_impl(graceful: bool) {
+    // form SQL connection
+    let conninfo = "host=127.0.0.1 port=5433 user=postgres replication=database";
+    let (client, connection) = tokio_postgres::connect(conninfo, NoTls).await.unwrap();
+    tokio::spawn(async move {
+        if let Err(e) = connection.await {
+            eprintln!("connection error: {}", e);
+        }
+    });
+
+    client
+        .simple_query("DROP TABLE IF EXISTS foo")
+        .await
+        .unwrap();
+    client
+        .simple_query("CREATE TABLE foo (i text)")
+        .await
+        .unwrap();
+
+    let slot = client
+        .simple_query("CREATE_REPLICATION_SLOT slot TEMPORARY LOGICAL \"test_decoding\"")
+        .await
+        .unwrap();
+
+    let lsn = match &slot[0] {
+        Row(row) => row.get("consistent_point").unwrap(),
+        _ => panic!("unexpeced query message"),
+    };
+
+    client
+        .simple_query("INSERT INTO foo VALUES ('fde61ce315faac78b17c')")
+        .await
+        .unwrap();
+
+    let query = format!("START_REPLICATION SLOT slot LOGICAL {}", lsn);
```

Comment:
Also test for an error case for `START_REPLICATION`, like an LSN that is way in the future or something. And right after the error, you can try a `close()`/`drop()` to make sure it still shuts down properly.

Can errors happen in the middle of the stream? I suppose so -- for instance, an OOM in `test_decoding`. I don't see an easy way to exercise this path, though, but if you have an idea it might be good to include that and make sure it doesn't desync the protocol.

Reply:
All error paths are now exercised :)